File: Online_Judge_Problems/boj/boj_17298_오큰수/boj_17298.py
"""
-----------------------------------------------------------
Sub    : [BOJ] 오큰수
Link   : https://www.acmicpc.net/problem/17298
Level  : Gold 4
Tag    : Python, Stack
-----------------------------------------------------------
Details

-----------------------------------------------------------
"""

import os
import sys

file_path = os.path.join(os.path.dirname(__file__), "input_test.txt")

if os.path.exists(file_path):
    f = open(file_path, "r", encoding="utf-8")
    input = f.readline
else:
    input = sys.stdin.readline


# Stack 클래스
class Stack:
    def __init__(self):
        self.items = []

    def push(self, data):
        self.items.append(data)

    def pop(self):
        return self.items.pop()

    def is_empty(self):
        return len(self.items) == 0

    def peek(self):
        return self.items[-1]


# input
N = int(input())
M = list(map(int, input().split()))

answer = [-1] * N
stack = Stack()

# solve
for i in range(N):
    while not stack.is_empty() and M[stack.peek()] < M[i]:
        index = stack.pop()
        answer[index] = M[i]

    stack.push(i)

# output
sys.stdout.write(" ".join(map(str, answer)))

File: boj/boj_17298_오큰수/boj_17298.js
/** * -----------------------------------------------------------
 * Sub    : [BOJ] 오큰수
 * Link   : https://www.acmicpc.net/problem/17298
 * Level  : 골드4
 * Tag    : JS, Stack
 * ------------------------------------------------------------
 * Details
 *
 * ------------------------------------------------------------
 */

const fs = require("fs");
const path = require("path");

const filePath = path.join(process.cwd(), "input_test.txt");
const input = fs.existsSync(filePath)
  ? fs.readFileSync(filePath, "utf8").trim().split(/\s+/)
  : fs.readFileSync(0, "utf8").trim().split(/\s+/);

let inputIdx = 0;

// Stack
class Stack {
  constructor() {
    this.items = [];
  }

  push(data) {
    this.items.push(data);
  }

  pop() {
    return this.items.pop();
  }

  peek() {
    return this.items[this.items.length - 1];
  }

  isEmpty() {
    return this.items.length === 0;
  }
}

// input & init
const N = Number(input[inputIdx++]);
const arr = [];
const stack = new Stack();
const result = Array(N).fill(-1);

for (let i = 0; i < N; i++) {
  arr[i] = Number(input[inputIdx++]);
}

// solve
for (let i = 0; i < N; i++) {
  while (!stack.isEmpty() && arr[stack.peek()] < arr[i]) {
    let index = stack.pop();
    result[index] = arr[i];
  }
  stack.push(i);
}

// output
process.stdout.write(result.join(" ") + "\n");
